File: Dynamic Programming/book_shop.py
INF = float('inf')
MOD = 10**9 + 7 
#TODO


def solve():
    n,x =  map(int,input().split())
    prices = list(map(int,input().split()))
    pages = list(map(int,input().split()))
    dp = [ [0 for i in range(x+1)] for i in range(len(prices) + 1) ]
    for i in range(1,n+1):
        for j in range(1,x+1):
            dp[i][j] = dp[i-1][j]
            if j - prices[i-1] >= 0:
                dp[i][j] = max(dp[i][j], dp[i-1][j - prices[i-1]] + pages[i-1])
    
    print(dp[n][x])




if __name__ == '__main__':
    solve()

File: Dynamic Programming/book_shop.cpp
#include <bits/stdc++.h>
using namespace std;
int main(){
 	int n,x;
 	cin >> n >> x;
 	vector<int> price(n), pages(n);
 	for(int i =0 ; i < n; i++) cin >> price[i];
 	for(int i = 0; i < n ; i++) cin >> pages[i];
	
 	vector< vector<int> > dp(n+1, vector<int>(x+1,0));
 	for(int i = 1; i <= n; i++){
 		for(int j = 1; j <= x; j++){
 			dp[i][j] = dp[i-1][j];
 			if(j - price[i-1] >= 0){
 				dp[i][j] = max(dp[i][j] , dp[i-1][j-price[i-1]] + pages[i-1]);
			 }
		 }
	 }
	 cout << dp[n][x] << endl;
}

